csfieldguide/static/interactives/dictionary-compression/js/dictionary-compression.js

```diff
@@ -0,0 +1,323 @@
+            var code_match = false;
+            var next_message_character_index = j + 1;
+            // if the character has already been matched, the move on to the next
+            if (message_characters[j].CodeIndex !== false) {
+                continue;
+            }
+            // if the first character of the code matches the current character in the message
+            if (code[0] == message_characters[j].Character) {
+                // check the rest of the code matches the following characters
+                if (code.length > 1) {
+                    for (var k = 1; k < code.length; k++) {
+                        // get the next code and message characters
+                        var next_code_character = code[k];
+                        var next_message_character = message_characters[next_message_character_index];
+                        // if a character in the message is already coded, move on
+                        if (next_message_character.CodeIndex != false) {
+                            break;
+                        }
+                        // if they match, then iterate to the next characters
+                        if (next_code_character == next_message_character.Character) {
+                            next_message_character_index += 1;
+                            // if they match and have reached the end of the code, then record that a match was found!
+                            if ((next_message_character_index - j) == code.length) {
+                                code_match = true;
+                            }
+                        } else {
+                            break;
+                        }
+                    }
+                } else { // if the code is one character long, then we know we have found a match
+                        code_match = true;
+                }
+            }
+            // before searching for the next occurance, mark the characters in the message
+            //     with which code index they correspond to
+            if (code_match == true) {
+                for (var l = 0; l < code.length; l++) {
+                    number_coded_characters += 1;
+                    var character_index = l + j;
+                    message_characters[character_index].CodeIndex = codes.indexOf(code);
+                    // interface
+                    var character_element = document.querySelectorAll('[data-character-index="' + character_index.toString() + '"]')[0];
+                    character_element.classList.add('compressed');
+                }
+            }
+        }
+    }
+    calculateStatistics(number_coded_characters);
+}
+
+function calculateStatistics(number_coded_characters) {
+    var percentage_coded = Math.round((number_coded_characters/message_characters.length) * 100);
+    var num_references_used = codes.length;
+    var stats = gettext('%(percentage_coded)% of text compressed using %(num_references_used) references.')
```

Comment:
Hi Hayley! Here to help you out with that translation bug. You almost had it!

Try something like this: 
```
d = { 
	'percentage_coded': percentage_coded,
	'num_references_used': num_references_used
};

var format = ngettext('%(percentage_coded)% of text compressed using 1 reference.', '%(percentage_coded)% of text compressed using %(num_references_used) references.', d.num_references_used);
s = interpolate(format, d, true);
```

This basically says if only 1 reference is used display the first string that is the non plural version, otherwise display the plural version. It decides if it should be plural or not based on num_references_used. This is why the third parameter is d.num_references_used. Interpolate passes in the variables to the format string and the 'true' boolean is saying that we should use named parameters.

Check this link for the official documentation :) https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#using-the-javascript-translation-catalog

Reply:
Does that fix the display problem though? Surely the problem is that a variable is being used in text that is translated before the JS is even compiled

One solution would be:
`stats = percentage_coded + '% ' + gettext('of text compressed using') + ' ' + num_references_used + ' ' + gettext('references.')`
Then what @courtneycb said would be used to make it look good if there's only one reference

Reply:
Nevermind, the docs say that is correct

Reply:
@courtneycb close, but not quite. My issue is that the first variable represents a number to then be used in a percentage, so `percent_coded` is an integer, and I want it to be rendered as `<percent_coded>%`. 
This:
```
    d = { 
        'percent_coded': percentage_coded,
        'num_refs_used': num_references_used
    };
    var format = ngettext('%(percent_coded) \% of text compressed using 1 reference.', '%(percent_coded) \% of text compressed using %(num_refs_used)s references.', d.num_refs_used);
    var s = interpolate(format, d, true);
    document.getElementById('interactive-dictionary-compression-stats').innerHTML = s;
```
Results in:
![Screenshot from 2019-04-10 18-53-30](https://user-images.githubusercontent.com/6937219/55873333-f5132880-5bc1-11e9-977d-8cdb19054193.png)

I've tried just adding a space between the number and the percentage sign and also tried usingt he escape character, neither work. It does work without the percentage sign, but then the statement doesn't make sense. Ideas?...

Reply:
> Does that fix the display problem though? Surely the problem is that a variable is being used in text that is translated before the JS is even compiled
> 
> One solution would be:
> `stats = percentage_coded + '% ' + gettext('of text compressed using') + ' ' + num_references_used + ' ' + gettext('references.')`
> Then what @courtneycb said would be used to make it look good if there's only one reference

FYI @eAlasdair we need to avoid creating strings this way as other languages can use a different word order to English.

Reply:
> @courtneycb close, but not quite. My issue is that the first variable represents a number to then be used in a percentage, so `percent_coded` is an integer, and I want it to be rendered as `<percent_coded>%`.
> This:
> 
> ```
>     d = { 
>         'percent_coded': percentage_coded,
>         'num_refs_used': num_references_used
>     };
>     var format = ngettext('%(percent_coded) \% of text compressed using 1 reference.', '%(percent_coded) \% of text compressed using %(num_refs_used)s references.', d.num_refs_used);
>     var s = interpolate(format, d, true);
>     document.getElementById('interactive-dictionary-compression-stats').innerHTML = s;
> ```
> Results in:
> ![Screenshot from 2019-04-10 18-53-30](https://user-images.githubusercontent.com/6937219/55873333-f5132880-5bc1-11e9-977d-8cdb19054193.png)
> 
> I've tried just adding a space between the number and the percentage sign and also tried usingt he escape character, neither work. It does work without the percentage sign, but then the statement doesn't make sense. Ideas?...

I think it's because you're missing the `s` after the brackets:

```
var format = ngettext('%(percent_coded)s \% of text compressed using 1 reference.', '%(percent_coded) \% of text compressed using %(num_refs_used)s references.', d.num_refs_used);
```

See: https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#interpolate

Reply:
> FYI @eAlasdair we need to avoid creating strings this way as other languages can use a different word order to English.

Yes I suspected so, I gave my best solution for the given sentence